### ISSUE TITLE
Feature Request: PJSIP VoIP call on API Trigger and/or VoIP virtual phone

### ISSUE THREAD

**Issue by TheBlueKingLP** [AlexxIT/go2rtc#552]
It would be awesome if this is able to initiate a 2 way audio + video call to a VoIP server such as Asterisk or FreePBX so I can receive calls from my Reolink doorbell or any other camera supporting 2 way audio, which already works great, with this.

`[Doorbell]  <---RTSP--->  [go2rtc] <---SIP Secure + Secure Real-time Transport Protocol---> [SIP Soft Phone or PBX]`

How it should work:

Plan A:
When an API endpoint is triggered, go2rtc should initiate a video VoIP call, bridging the video feed from the camera to the call. i.e. go2rtc should place a VoIP call to a specified SIP VoIP server or [SIP URI](https://www.3cx.com/pbx/sip-uri/)

Plan B:
go2rtc should act as a SIPS client(i.e. a virtual video phone), when it receives a call, it should bridge the camera video feed, microphone and audio from the camera, to the video call.
i.e. go2rtc should accept the VoIP call and act like a SIP phone with video capability.

**Comment by AlexxIT** [comment]
- why two plans?
- what is endpoint? Go2rtc special new API for init call?
- what will be after Asteriks get the call?

**Comment by TheBlueKingLP** [comment]
1. 2 plan to see which one people prefers and/or easier to implement
2. as long as it can be triggered, it is fine
3. sorry, not quite sure what you mean here

**Comment by AlexxIT** [comment]
I'm not have experience with SIP technology. I only know about SDP exchange process on start. I don't understand why people still use this technology.

**Comment by TheBlueKingLP** [comment]
It is the industry standard for VoIP phones. As far as I know all VoIP phones uses this protocol and I haven't heard of one that uses something else. It should be simple to implement with something like [srtp](https://github.com/pion/srtp) and [sip](https://github.com/ghettovoice/gosip) libraries. Haven't looked at those in depth but they might work?

**Comment by AlexxIT** [comment]
So you have some special device (phone), that can accept video call from Asterisk?

I can't understand why anyone would need special phones when everyone has a smartphone.

**Comment by TheBlueKingLP** [comment]
It does not have to be a special device/phone. There are software phones for android/iOS/macOS/Windows/Linux for example [Linphone](https://linphone.org/)(free and open source) or [Bria](https://www.counterpath.com/bria-solo/) (paid or free trial). Or calls can be redirected to a public phone line by using a SIP trunk(a phone line that connects VoIP to the public switched telephone network). For example [Twilio voice](https://www.twilio.com/en-us/voice) or [VoIP.ms](https://voip.ms/residential)

**Comment by AlexxIT** [comment]
What the easiest and friendly way for setup SIP server for tests?

**Comment by TheBlueKingLP** [comment]
FreePBX. [Complete setup guide](https://www.youtube.com/watch?v=l30WKTYf9ZY&list=PL1fn6oC5ndU_umAhL9A_1zkC90hMPDPNO)

**Comment by backcountrymountains** [comment]
This is also something I'm interested in. Just having the doorbell call my phone and have a phone conversation with whoever is at the door.

Really, the dream is to be able to have a video call, like facetime or google meet/duo, whoever is at the door and have a tablet that displays my dumb face to the person at the door (but only if I want them to see me).

Anyway, +1 vote for SIP calling doorbell.

**Comment by TheBlueKingLP** [comment]
> This is also something I'm interested in. Just having the doorbell call my phone and have a phone conversation with whoever is at the door.
> 
> Really, the dream is to be able to have a video call, like facetime or google meet/duo, whoever is at the door and have a tablet that displays my dumb face to the person at the door (but only if I want them to see me).
> 
> Anyway, +1 vote for SIP calling doorbell.

SIPS + SRTP **does** support video call, only one way though(you will be able to see the person who is at the door), unless you have someway(a wall display?) to display your face to them

**Comment by AlexxIT** [comment]
I don't know if it possible to use facetime API.
I think it's possible to use Duo API. But any Google API always crazy hard to use.
I think it's possible to use Telegram API. And it shouldn't be very hard. But call should be done from account of some exist user.

**Comment by TheBlueKingLP** [comment]
I would prefer SIPS + SRTP though, this allows local offline use and selfhost while the ones you have mentioned requires cloud.

**Comment by felipecrs** [comment]
This is related to https://github.com/AlexxIT/go2rtc/issues/409

**Comment by Vaarlion** [comment]
Also interested because i could integrate the Reolink doorbell with any smart display and android tablet without the need for a cloud based application.

**Comment by los93sol** [comment]
This would be awesome as it would effectively enable SIP support for any camera that go2rtc can support 2 way audio for.  I agree on the surface it seems silly, but it actually does enable a lot of options for people to automate.

**Comment by felipecrs** [comment]
> What the easiest and friendly way for setup SIP server for tests?

You may also find our Asterisk add-on to come in handy: https://github.com/TECH7Fox/asterisk-hass-addons#readme

I use it daily for a fully local and HA communication with my Dahua VTO doorbell.

**Comment by los93sol** [comment]
@felipecrs I'm actually using it already, but I'm stuck with how to get my cameras working with it since they don't support SIP.  I'm able to get the sip card to show the camera, but not exactly what I'm looking for because two way audio won't work with that arrangement.

**Comment by felipecrs** [comment]
lol if your cameras don't support SIP why are you even using it then? just use go2rtc with the webrtc card or the frigate card.

**Comment by los93sol** [comment]
I have that, but what I really want is to receive a call which having it connected to asterisk would enable for me

**Comment by felipecrs** [comment]
I see now, you want to convert a non-SIP camera to SIP basically.

**Comment by los93sol** [comment]
Exactly, thought I could get away with using rtsp-app in the asterisk addon to pull it off, but doesn't look like it's compiled into the version for the Pi so got walled there too.  If you've got any ideas how to do it please lmk

**Comment by felipecrs** [comment]
If you mean https://github.com/tommyjlong/app_rtsp_sip, yes, it is compiled into the add-on, for all platforms.

If you want to continue this discussion, please open an issue in the add-on. I don't think people subscribed here wants to be part of this conversation.
